.github/workflows/docker-build-push-multiarch.yml

```diff
@@ -0,0 +1,367 @@
+          # runner labels. It's returned in the form of a JSON object
+          # so it can be passed directly into a matrix strategy.
+          #
+          # Ex: If PLATFORMS=linux/arm64,linux/amd64
+          #     then MATRIX=[ "arm64","x64" ]
+          #############################################################
+
+          # Create a copy so we don't override our input
+          # This runs afoul of spell check, which thinks PLATFORMS should be PLATFORM
+          # shellcheck disable=SC2153
+          PS=$PLATFORMS
+
+          # The second half of this command converts $ARCH_TO_PLATFORM_MAP into string pairs of ARCH and PLATFORM
+          # The first half of this command loops over those items and string replaces PLATFORM with ARCH
+          # In the above example, the end result of this will be: PS=arm64,x64
+          while read -r ARCH PLATFORM; do PS=${PS//$PLATFORM/$ARCH}; done < <(echo "$ARCH_TO_PLATFORM_MAP" | jq -r 'to_entries[] | "\(.key) \(.value)"')
+
+          # Convert to single-line JSON array
+          MATRIX=$(echo "$PS" | jq -R -c 'split(",")')
+
+          # Export as GitHub Action output
+          echo "runner_arches=$MATRIX" | tee -a "${GITHUB_OUTPUT}"
+
+  build-and-push:
+    needs: prepare-matrix
+    strategy:
+      fail-fast: false
+      matrix:
+        arch: ${{ fromJson(needs.prepare-matrix.outputs.runner_arches) }}
+    runs-on: ubuntu-${{ matrix.arch }}-${{ inputs.server-size }}
+    outputs:
+      annotations: ${{ steps.build.outputs.annotations }}
+      digest: ${{ steps.build.outputs.digest }}
+      imageid: ${{ steps.build.outputs.imageid }}
+      images: ${{ steps.build.outputs.images }}
+      json: ${{ steps.build.outputs.json }}
+      labels: ${{ steps.build.outputs.labels }}
+      metadata: ${{ steps.build.outputs.metadata }}
+      tags: ${{ steps.build.outputs.tags }}
+      version: ${{ steps.build.outputs.version }}
+    permissions:
+      contents: read
+      id-token: write
+    steps:
+      - name: Prepare
+        run: |
+          #############################################################
+          # Take incoming arch and lookup the matching docker style platform
+          # Then create a unique platform pair, to use for our digests later
+          #
+          # Ex: if arch=x64 then PLATFORM=linux/amd64 and PLATFORM_PAIR=linux-amd64
+          #############################################################
+
+          PLATFORM=$(echo "$ARCH_TO_PLATFORM_MAP" | jq -r --arg arch "${ARCH}" '.[$arch]')
+          echo "PLATFORM=${PLATFORM}" | tee -a "${GITHUB_ENV}"
+          echo "PLATFORM_PAIR=${PLATFORM//\//-}" | tee -a "${GITHUB_ENV}"
+        env:
+          ARCH: ${{ matrix.arch }}
+      - name: Checkout code
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+        with:
+          persist-credentials: false
+      - name: Run custom build script if it exists
```

Comment:
Just a bit of feedback -- take it or leave it 🙂

I would consider leaving the scripting components out of the "boxed up" reusable workflow. I think it's OK enough to say that if you need to run code before/after your build then you can use the "parts-based" workflow. It gives people the flexibility to do more than you can do with a script too, for example to use third party actions.